### PR TITLE
perf(octane): bring JSX keyed-list rendering closer to TSRX

### PR DESCRIPTION
**Pull request by domgan-oai**
## Summary

- Bring compiled TSX/JSX keyed-list updates substantially closer to TSRX without assuming an arbitrary `.map` is `Array.prototype.map`.
- Reduce unchanged 1,000-row JSX updates from approximately **0.163 ms to 0.00122 ms** (~**134x faster**) and single-row updates from approximately **0.166 ms to 0.046 ms** (~**3.6x faster**).
- Stack cleanly on #361 at `9a5c0e2d42e0afb486ebd8d8be52ae1f088abb90`, including its scoped JSX, deep-SSR, lazy child-accessor, dynamic-fragment, and i18next compatibility fixes.

## Implementation

- Enable purity and automatic memoization analysis for eligible returned JSX; share TSRX's keyed-list reconciliation and per-item reuse while preserving JSX's normal public descriptor boundary.
- Guard the optimized path to genuine dense native Arrays and the original intrinsic `Array.prototype.map`; evaluate the receiver and method once, then dispatch exotic/custom/overridden/subclass/species/sparse/accessor cases through the original method with its original receiver.
- Preserve callback indices, indexed getters, constructor/`Symbol.species` effects, lexical `this`/`arguments`, context invalidation, server rendering, hydration identity, foreign DOM ownership, and transitions between native and custom receivers.
- Add exhaustive development/production compiler, runtime, server, hydration, and integration regressions, deterministic JSX/TSRX compiled-work assertions, stronger production benchmark ceilings, and an Octane patch changeset.

## Validation

- Focused integrated scoped JSX, exotic maps, directive lowering, deep Fizz, foreign/portal ownership, Docusaurus, i18next, and i18next SSR: **23 files / 865 tests passed**.
- Full integrated suite across `octane`, `octane-prod`, `docusaurus`, `i18next`, and `i18next-ssr`: **790 files / 10,322 tests passed** (`./node_modules/.bin/vitest run --project octane --project octane-prod --project docusaurus --project i18next --project i18next-ssr --reporter=dot`).
- `node node_modules/prettier/bin/prettier.cjs --check .`
- `./node_modules/.bin/tsgo --noEmit -p packages/octane/tsconfig.json`
- `./node_modules/.bin/tsgo --noEmit -p packages/octane/typetests/tsconfig.json`
- `node scripts/check-changesets.js`
- `node scripts/check-test-markers.mjs`
- `node scripts/check-tsrx-module-decls.mjs`
- JSX/TSRX normalized gzip ceilings remain intact: application **2882 / 2402 = 1.19983x < 1.2x**, framework **33820 / 17975 = 1.88153x < 1.9x**, and total **36702 / 20377 = 1.80115x < 1.82x**.
- Fresh production Chromium comparisons pass the tightened JSX/TSRX guards: unchanged **5.17x < 6x** (previous ceiling 600x), one-row change **7.57x < 10x** (previously 32x), context fanout **2.19x < 2.25x** (previously 3x), and mount **1.34x < 1.6x**. Deterministic work checks require just **three JSX descriptors and zero keyed-survivor visits** on unchanged updates, and exactly one row/item body on single-row updates.
- Direct-host SSR remains byte-identical for 1,000 rows: **29,817 raw bytes / 4,865 gzip bytes / exactly two outer markers**, with server node identity retained during hydration and reordering.
- Existing aggregate codegen-size ceiling is already red on the unchanged #361 base: **26,035 / 24,233 = 1.07436x > 1.04x**; this branch measures **26,175 / 24,233 = 1.08014x** (+140 gzip bytes in the JSX corpus, with all 15 TSRX fixtures byte-identical). No existing ceiling was weakened.

## Integration

Depends on #361. Its commits currently appear in this PR while #361 is open; once #361 merges, the comparison collapses to this single focused performance commit. Preserve #361's directive-aware return lowering, lazy scoped descriptors, dynamic fragment classifier/i18next compatibility, and deep-SSR replay behavior when integrating.

---

> [!NOTE]
> **High Risk**
> Changes core compiler lowering and client/server reconciliation for JSX lists; incorrect guarding or keyed reuse would cause subtle render bugs, though behavior is heavily regression-tested.
> 
> **Overview**
> **Returned JSX keyed lists** now participate in the same purity, whole-list autoMemo, and per-key survivor reuse as TSRX `@for`, while keeping the normal return-descriptor wrapper and JSX children semantics.
> 
> Eligible `{items.map(...)}` lowers through a **guarded** synthetic loop: receiver and `map` are evaluated once, then **`mapSlot`** checks for a dense native `Array.prototype.map` before emitting compiled keyed reconciliation; exotic receivers, sparse arrays, extra arguments, lexical `this`/`arguments` in callbacks, and overridden `map` stay on ordinary dispatch. Return-JSX components get component-local capture analysis so list regions can be memoized; SSR mirrors the same branch via `mapSlot` + `ssrChild`.
> 
> **Runtime** adds client/server `mapSlot`, extends `childSlot` for compiled map bodies (including native↔custom transitions, component rows, hydration adoption), and small scoped-children fixes for providers.
> 
> **Benchmarks/docs:** memo-wall README and `work.mjs` gain `WORK_DIALECT=jsx`; wall-B expectations reflect cached helper output; `ratios.json` tightens JSX vs TSRX ceilings (e.g. equal parent rerender **600→6**). Large **`auto-memo.test.ts`** coverage for maps, hydration, and edge cases. Patch changeset for `octane`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit bff72b4ce851ef9924c8765c76b56e6bfd40f24f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/octane/src/runtime.ts [hunk 1]
```diff
@@ -16142,16 +16429,139 @@ export function childSlot(
...
+										(text as any).$$deoptKey = 0;
+										break;
+									}
+								}
+							}
+						}
+					}
+					block.deoptNode = block.startMarker;
+				}
+			}
```
[the remark this answers is not in the document]
Comment (domgan-oai) [reply packages/octane/src/runtime.ts:16645]: Fixed in 41ca89ca. Native ↔ custom `.map()` transitions now preserve keyed component/DOM identity, component state and context, and correctly reconcile occupied component slots in both directions. The follow-up also fixes native callback ordering after custom reordering, hydration, mixed component/host/null results, and stale effect cleanup/remounting.

Added 10 targeted regressions: 88 focused development/production cases and the complete 10,332-test suite pass, with the existing performance and gzip limits unchanged.
